app/src/main/java/io/apicurio/registry/ibmcompat/api/impl/ApiServiceImpl.java

```diff
@@ -52,12 +77,147 @@
+                      .collect(Collectors.toList());
+    }
+
+    private SchemaVersion getLatestSchemaVersion(String schemaid) {
+        return getSchemaVersion(schemaid, storage.getArtifact(schemaid).getVersion());
+    }
+
+    private SchemaVersion getSchemaVersion(String schemaid, Long versionid) {
+        SchemaVersion schemaVersion = new SchemaVersion();
+        try {
+            ArtifactVersionMetaDataDto artifactMetaData = storage.getArtifactVersionMetaData(schemaid, versionid);
+            schemaVersion.setId(versionid.intValue()); // TODO not safe!
+            schemaVersion.setDate(new Date(artifactMetaData.getCreatedOn()));
+            schemaVersion.setName(artifactMetaData.getName());
+            schemaVersion.setEnabled(!ArtifactState.DISABLED.equals(artifactMetaData.getState()));
+            SchemaState versionState = new SchemaState();
+            if(ArtifactState.DEPRECATED.equals(artifactMetaData.getState())) {
+                versionState.setState(SchemaState.StateEnum.DEPRECATED);
+            } else {
+                versionState.setState(SchemaState.StateEnum.ACTIVE);
+            }
+            schemaVersion.setState(versionState);
+        } catch (ArtifactNotFoundException e) {
+            return null;
+        }
+        return schemaVersion;
+    }
+
+    private void populateSchemaSummary(String schemaid, SchemaSummary schemaSummary) {
+        List<ArtifactState> versionStates = storage.getArtifactVersions(schemaid).stream()
+            .map(version -> storage.getArtifactVersionMetaData(schemaid, version).getState())
+            .collect(Collectors.toList());
+
+        schemaSummary.setId(schemaid);
+        schemaSummary.setName(schemaid); // TODO - add mechanism to store an artifact-level name
+
+        // The schema is disabled if all versions are disabled
+        boolean isSchemaDisabled = versionStates.stream().allMatch(versionState -> ArtifactState.DISABLED.equals(versionState));
+        schemaSummary.setEnabled(!isSchemaDisabled);
+
+        // The schema is deprecated if all versions are deprecated
+        boolean isSchemaDeprecated = versionStates.stream().allMatch(versionState -> ArtifactState.DEPRECATED.equals(versionState));
+        SchemaState schemaState = new SchemaState();
+        if(isSchemaDeprecated) {
+            schemaState.setState(SchemaState.StateEnum.DEPRECATED);
+        } else {
+            schemaState.setState(SchemaState.StateEnum.ACTIVE);
+        }
+        schemaSummary.setState(schemaState);
+    }
+
+    private void handleArtifactCreation(AsyncResponse response, String artifactId, String versionName, Throwable t) {
+        if (t != null) {
+            response.resume(t);
+            return;
+        }
+
+        if (!doesArtifactExist(artifactId)) {
+            // Some storage (such as asyncmem) needs a wait before the artifact is available for updating
+            try {
+                waitForArtifactCreation(artifactId);
```

Comment:
This seems like an OK short term solution.  But it would be good to think about what other options we might have here to avoid this.  Especially since the current implementation is hard coded to wait for about 1s.  Any async storage (infinispan and kafka streams are both async) cannot make a 1s guarantee.  So we would at least need to make the wait time adjustable.  But even better would be if we could enhance the storage layer in some way to avoid this wait time.  I'm open to ideas!  :)

Reply:
Any kind of wait / retry should be dealt differently -- see how we handle this with AsyncBiFunctionService in Streams storage.
So you actually get a remote callback.

Reply:
Yep - the reason for this block is due to how AsyncInMem storage adds a delay to the creation timestamp for the new artifact [1] and so when I call the `updateArtifactMetaData()` call (at line 157, which is currently commented out), it checks the artifact creation time, and returns an `ArtifactNotFoundException` as the creation timestamp has not yet been met, due to the added delay [2]. 

[1] https://github.com/Apicurio/apicurio-registry/blob/54d82c269c6f57be4f80c7bae4a9b00dfe789ffd/storage/asyncmem/src/main/java/io/apicurio/registry/asyncmem/AsyncInMemoryRegistryStorage.java#L105-L107

[2] https://github.com/ajborley/apicurio-registry/blob/54d82c269c6f57be4f80c7bae4a9b00dfe789ffd/storage/asyncmem/src/main/java/io/apicurio/registry/asyncmem/AsyncInMemoryRegistryStorage.java#L360-L363

Potential fixes for this could be to allow some of the metadata to be set as part of the artifact creation, or maybe somehow ignore the timestamp check when making `updateArtifactMetaData` calls within the server.

I haven't seen similar issues with other storage types needing a wait/retry, but I haven't checked them all extensively.

Reply:
The purpose of the `asyncmem` storage type is to mimic the async nature of the Kafka based storages.  Especially when in a clustered environment, the Kafka storages may return a 404 for a GET immediately after a POST (create an artifact and then quickly ask for it may result in a NotFound).  This is from a remote client perspective, of course.  We had some issues in our integration tests around this - the tests need to understand that a slight delay is possible.  So I created the `asyncmem` storage to force the delay so I could shake out any problems in the tests.

(Note:  the integration tests are run by QE against real deployments, including clustered deployments)

So I think this is somewhat of an unintended consequence.  Still, I think we should open a separate GitHub issue to discuss the things you're trying to do in the ibmcompat API layer - we can enhance the storage interface to accommodate those needs.